### PR TITLE
Feature/refactorize framework proto

### PR DESCRIPTION
**Pull request by reyoung**
1. Update to develop branch.
2. Change `op_proto_name` -> `parameter`, `var_names` -> `arguments`
  * commit b368c6c
3. Rewrite intermediate representation in Operator and Network.
  * We use a special attribute `temporary_index` to describe which outputs variable is intermediate.
  * Now we have two situations.
     * For Operator, the intermediate output's parameter name is marked in `OpProto`. Just read OpProto to get them.
     * For NetOp, the intermediate variables are calculated in `NetOp::CompleteAddOp()` method, and store in a set named `NetOp::intermediate_outputs_`;
     * Add `OperatorBase::OutputVars` to get outputs with or without intermediate_outputs.
     * commit 78af6e601

This PR make all C++ unit tests passed except

* Backward
* RNN (which is done by #5 )

All Python unit tests is under fixing.

### CODE REVIEW

File: paddle/framework/op_registry.h [hunk 1]
```diff
@@ -127,25 +128,14 @@ class OpRegistry {
...
+    OpProto& op_proto = OpProtos()[op_type];
     auto maker = ProtoMakerType(&op_proto, &op_checker);
     maker.Validate();
     *op_proto.mutable_type() = op_type;
     PADDLE_ENFORCE(
         op_proto.IsInitialized(),
         "Fail to initialize %s's OpProto, because %s is not initialized",
         op_type, op_proto.InitializationErrorString());
-
-    VarIndexMaps()[op_type].reset(new VarIndexMap());
```
Comment (reyoung) [inline paddle/framework/op_registry.h:139]: VarIndexMap以前被用来表示OpDesc中Parameter和Arguments的关系，目前已经不需要了。

File: paddle/framework/op_registry.h [hunk 2]
```diff
@@ -212,22 +202,11 @@ class OpRegistry {
     return grad_op;
   }
 
-  static std::unordered_map<std::string, OpProto>& protos() {
```
Comment (reyoung) [inline paddle/framework/op_registry.h:215]: 将 protos() 移动到Operator.h中，让OperatorBase也可以获得到OpProtos，避免交叉依赖。

现在的依赖关系是 op_registry依赖operator.

File: paddle/framework/operator.cc [hunk 3]
```diff
@@ -33,6 +33,14 @@ ExecutionContext::GetEigenDevice<platform::GPUPlace, Eigen::GpuDevice>() const {
 }
 #endif
 
+static std::unordered_map<std::string, OpProto>* g_op_protos = nullptr;
```
Comment (reyoung) [inline paddle/framework/operator.cc:36]: 这里同时把原来的函数内静态变量改成了全局静态变量指针。更符合google code style
https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

**Comment by wangkuiyi** [comment]
It seems that this PR conflicts with the target branch after I merged @qingqing's PR.  Let me add @reyoung and @qingqing as an collaborator, so you could merge in your will.

**Pull request merged**